### PR TITLE
Java Constant Converter

### PR DESCRIPTION
**Pull request by kennyqiyenkan**
Users can now use the `constants` to add constants to their converted java code.

### CODE REVIEW

File: src/foam/java/Class.js [hunk 1]
```diff
@@ -100,6 +112,11 @@ foam.CLASS({
       return this;
     },
 
+    function constant(c) {
+      this.constants.push(foam.java.constant.create(c));
```
Comment (kgrgreer) [inline src/foam/java/Class.js]: Should this be foam.java.Constant? How is this working?

File: src/foam/java/Constant.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.java',
+  name: 'Constant',
+
+  properties: [
+    'name',
+    'javaType',
```
Comment (kgrgreer) [inline src/foam/java/Constant.js]: This is already java specific, so we just need 'type' with java being implied.
Also, we only need 'value', which is the 'javaValue'.

File: src/foam/java/Constant.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    'documentation'
+  ],
+
+  methods: [
+    function outputJava(o) {
+      o.indent();
+
+      if ( this.documentation != undefined ) {
+        o.out('/* ', this.documentation, ' */\n');
+        o.indent();
```
Comment (kgrgreer) [inline src/foam/java/Constant.js:35]: What are the next two lines for?

File: src/foam/java/Constant.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        o.indent();
+      }
+
+      o.out('public static final ');
+
+      if ( this.type == undefined ) {
```
Comment (kgrgreer) [inline src/foam/java/Constant.js]: This logic should be in the factory for the 'type' property.

File: src/foam/java/Constant.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+      }
+
+      o.out(' ' + this.name);
+      o.out(' = ');
+
+      // if the user has declared a type that requires a custom instantiation,
+      // but not declare a javaType for it, the Java code should break on compile
+      if ( this.javaValue == undefined ) {
+        if ( foam.String.isInstance(this.value) ) {
+          o.out('\"' + this.value + '\"');
```
Comment (kgrgreer) [inline src/foam/java/Constant.js]: You'll need to escape this.value in case it contains "'s.

File: src/foam/java/Constant.js (hunk 3 above)
Comment (kennyqiyenkan) [reply src/foam/java/Constant.js:35]: the tabbing doesn't come out well after the `\n`. The two indents fixes that.

File: src/foam/java/Constant.js (hunk 5 above)
Comment (kennyqiyenkan) [reply src/foam/java/Constant.js]: Should that not be handled by the user?

File: src/foam/java/Constant.js (hunk 4 above)
Comment (kennyqiyenkan) [reply src/foam/java/Constant.js]: what would you mean by that? As in we do not need to check if the type is undefined?

File: src/foam/java/Constant.js (hunk 2 above)
Comment (kennyqiyenkan) [reply src/foam/java/Constant.js]: type is understandable.

as for javaValue/value, how would we be able to discern if they have their own custom instantiation or they are simply using a primitive?

File: src/foam/java/Constant.js (hunk 5 above)
Comment (kgrgreer) [reply src/foam/java/Constant.js]: The user could do something like:
constants: { QUOTE_CHARACTER: '"' } OR
constants: { QUOTE_CHARACTER: "\"" }

Which you would output as """. How could the user handle this?

Comment (kennyqiyenkan) [reply src/foam/java/Constant.js]: mm.. I believe I see your point.

**Pull request merged**